### PR TITLE
[WIP] LD Local Links

### PR DESCRIPTION
**Pull request by zwrawr**
# LD local links
Make local links look good and not cause reloads
image

Somewhere in the gitter chat mentions how `//` and `///` Links are meant to work but i'll have to look back to find out. Leaving this for a while till I get more time.

## TODO 
* Get `//` links working. (I think these should work the same as `/` links)
* Get `///` links working. Linking to images, but not as an image.
* Get relative links working
* Remove other logging

### CODE REVIEW

**Comment by local-minimum** [comment]
This is probably just reflecting my understanding but why is there need for the extra uri-logic in the autoembed.js? I kind of though that what cont'd out of the marked render should be ready to use?

**Comment by zwrawr** [comment]
Autoembed.js handles site specific embeds. E.g. youtube players or twitter embeds.

In this case its being used for ld specific link styling. The way were detecting internal links is kinda broken so i still need to clean it up.

**Comment by local-minimum** [comment]
Yeah I think that the way there was stuff about links being internal there or not was confusing me. 

Anyhow , this relates to https://github.com/ludumdare/ludumdare/issues/879 I think.

**Comment by povrazor** [comment]
Basic fix to solve merge conflicts. May trigger the linter since this code was previously repaired.

File: src/com/autoembed/autoembed.js [hunk 1]
```diff
@@ -47,12 +53,20 @@ export default class AutoEmbed extends Component {
     }
 
     if (url.domain) {
+
+      /*
+      * Embeded content
+      */
       if (url.domain.indexOf('youtube.com') !== -1) {
         if ((url.paths && url.paths[0] == 'watch') && url.args.v) {
           return <YoutubeEmbed href={props.href} />;
```
Comment (povrazor) [inline src/com/autoembed/autoembed.js]: I realize this isn't from this PR (again, I'm taking notes for linter/style guide improvements), but prefer this:
```
if ( url.domain.indexOf('youtube.com') !== -1 ) {
          if ( url.paths && (url.paths[0] == 'watch') && url.args.v ) {
```
I mention spacing of special keywords elsewhere.

Of course use added brackets as needed, but a rule of thumb is to look at what's inside the conditional. 

* If it's a single conditional/implicit statement (i.e. `if ( url.paths && url.args.v ) {`) it's fine without added brackets
* If it's only a single `LHS == RHS` statement, it's fine without added brackets
* If multiple statements are used, separated by `&&` or `||`, and one is a `LHS == RHS` statement, use added brackets just to visually isolate it. i.e. `if ( other && (LHS == RHS) ) {`
* If the LHS or RHS side is more than a single item, use brackets to visually isolate the complexity. i.e. `if ( (a+b) == 12 ) {`

`&&` operations in series will always require the previous one to succeed, so no extra brackets are needed. When you mix and match `&&` with `||`, extra brackets will be necessary to guarantee order.

`&&`, `||`, assignment (`=`), and equality tests (`==`, `!==`, `>=`, etc) are special, and should be given extra spaces around them to make them easier to notice. 

Whether to space arithmetic operators (`+`, `-`, etc) is up to the person writing the math. If it's easier to follow with spaces, then use spaces.

File: src/com/autoembed/autoembed.js [hunk 2]
```diff
@@ -77,9 +91,22 @@ export default class AutoEmbed extends Component {
         return (
           <SmartLink icon_name='soundcloud' full_url={url.href} domain={lit} part_url={unlit}></SmartLink>
         );
-      } else if ( url.domain.indexOf('//'+window.location.hostname) !== -1 ) {
-        return (<span class="smart-link"><a href={props.href}><strong class="-the-rest">{props.href}</strong></a></span>);
-			}
+      } else if (url.domain.indexOf(window.location.hostname) == 0 ) {
```
Comment (povrazor) [inline src/com/autoembed/autoembed.js]: Prefer
```
}
else if ( ... ) {
```

File: src/com/autoembed/autoembed.js [hunk 3]
```diff
@@ -77,9 +91,22 @@ export default class AutoEmbed extends Component {
...
+        // local links starting pointing to the same domain name e.g. http://ldjam.com/users/mike
+        console.log(lit);
+        console.log(unlit);
+        console.log(url);
+        console.log(props);
```
Comment (povrazor) [inline src/com/autoembed/autoembed.js]: Try to remember to comment out or remove logging that is no longer necessary.

File: src/com/autoembed/autoembed.js [hunk 4]
```diff
@@ -77,9 +91,22 @@ export default class AutoEmbed extends Component {
...
+
+        return (
+          <LocalLink href={url.href} text={(unlit)?url.path:props.text} title={''} target={"_blank"}/>
+        );
+      } else {
```
Comment (povrazor) [inline src/com/autoembed/autoembed.js]: Prefer
```
}
else {
```

File: src/com/autoembed/locallink.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+import NavLink        from 'com/nav-link/link';
+import SVGIcon        from 'com/svg-icon/icon';
+
+export default class LocalLink extends Component {
+  constructor(props) {
+    super(props);
+  }
+
+  render(props, {}) {
```
Comment (povrazor) [inline src/com/autoembed/locallink.js]: Prefer
```
render( props, {} ) {
```

File: src/com/autoembed/smartlink.less [hunk 6]
```diff
@@ -13,3 +15,31 @@
...
+                color: @COL_A;
+            }
+            &.icon-d-are{
+                color: @COL_AB;
+            }
+        }
+    }
+
+    &:hover,
+    &:active{
```
Comment (povrazor) [inline src/com/autoembed/smartlink.less]: Prefer
```
&:active {
```
(notice the space between active and `{`)

File: src/internal/autoembed/autoembed.js [hunk 7]
```diff
@@ -40,20 +40,19 @@
...
 
-		// if its not already an external link with a protocol and it has a dot in it
-		// then make it an external link becuase not internal links have dots
-		if ( (ret.url.indexOf('//') == -1) && (ret.url.indexOf('.') != -1) && (ret.url.indexOf('/') == -1 || ret.url.indexOf('.') < ret.url.indexOf('/')) ) {
-			// Assume HTTP. Most HTTPS sites will autoredirect to HTTPS
-			ret.url = 'http://' + ret.url;
-		}
-		else if ( ret.url.indexOf('//') == -1 ) {
-			ret.internal = true;
+		if (ret.origin.indexOf('/') == 0 || ret.origin.indexOf('.') == -1) {
```
Comment (povrazor) [inline src/internal/autoembed/autoembed.js]: Prefer 
```
if ( (ret.origin.indexOf('/') == 0) || (ret.origin.indexOf('.') == -1) ) {
```

File: src/internal/marked/Renderer.js [hunk 8]
```diff
@@ -4,14 +4,193 @@ import Util 				from './Util';
...
 export default class Renderer {
-	constructor(options) {
-		this.options = options || {};
-	}
+  constructor(options) {
+    this.options = options || {};
+  }
+
+  code(code, lang, escaped) {
+    if (this.options.highlight) {
```
Comment (povrazor) [inline src/internal/marked/Renderer.js]: Prefer `if ( ... ) {` here and below

File: src/internal/marked/Renderer.js [hunk 9]
```diff
@@ -4,14 +4,193 @@ import Util 				from './Util';
...
+    return (
+      <blockquote>{quote}</blockquote>
+    );
+  };
+
+  html(html) {
+    return {html};
+  };
+
+  heading(text, level, raw) {
```
Comment (povrazor) [inline src/internal/marked/Renderer.js]: prefer
```
heading( text, level, raw ) {
```
'cause it's a function definition

File: src/internal/marked/Renderer.js [hunk 10]
```diff
@@ -4,14 +4,193 @@ import Util 				from './Util';
...
+      <BlockSpoiler>{secret}</BlockSpoiler>
+    );
+  };
+  blockquote(quote) {
+    return (
+      <blockquote>{quote}</blockquote>
+    );
+  };
+
+  html(html) {
```
Comment (povrazor) [inline src/internal/marked/Renderer.js]: prefer
```
html( html ) {
```
because it's a function definition

**Comment by zwrawr** [comment]
Closed in favour of #1293

**Pull request closed without merging**